crates/accelerate/src/quantum_circuit/circuit_data.rs

```diff
@@ -300,6 +306,155 @@ impl CircuitData {
         self.data.reserve(additional);
     }
 
+    /// Returns a tuple of the sets of :class:`.Qubit` and :class:`.Clbit` instances
+    /// that appear in at least one instruction's bit lists.
+    pub fn active_bits(&self, py: Python<'_>) -> Py<PyTuple> {
+        let mut qubits: HashSet<BitType> = HashSet::new();
+        let mut clbits: HashSet<BitType> = HashSet::new();
+        for inst in self.data.iter() {
+            for b in self.intern_context.lookup(inst.qubits_id).iter() {
+                qubits.insert(*b);
+            }
+            for b in self.intern_context.lookup(inst.clbits_id).iter() {
+                clbits.insert(*b);
+            }
+        }
+
+        (
+            PyList::new(
+                py,
+                qubits
+                    .into_iter()
+                    .map(|b| self.qubits_native[b as usize].clone_ref(py)),
+            ),
+            PyList::new(
+                py,
+                clbits
+                    .into_iter()
+                    .map(|b| self.clbits_native[b as usize].clone_ref(py)),
+            ),
+        )
+            .into_py(py)
+    }
```

Comment:
I switched to using `IndexSet` instead to get stability based on insertion order, which would be dictated by the order the bits first appear in `data`.

If we were to return a Python set, we'd still have the issue of instability I believe, since set is not ordered, correct?

Done in d42b4f74987f65f0c5afe09daf686ebf59f13bb9

Reply:
Sorry, I wasn't totally clear: I'm fine with an unordered return if we use an explicitly unordered collection as the return value (Python's `set`, for example).  Where I had a problem was that we were returning an implicitly ordered collection (`list`) that didn't actually have an order.  That's the kind of thing that's led to bugs for us before - it's very easy to think that `list` will always be ordered, but if you're working with a `set`, you know for sure that it's not.

I think `set` is probably the better return type, just thinking about how it's likely to be used; there will be cases where we'll want to iterate through all active qubits, but also commonly, I think people will want to get the set of active qubits, then query that collection for "is this object active?".  Both of those can be satisfied efficiently if we return `set`, but if we return `list`, then the second use-case ended up requiring two intermediate data structures that are immediately thrown away (the inner Rust set and the Python list).

Reply:
Ah I see. For the only current use-case, it's ideal to have these returned in a deterministic order, but I agree with you that in principle this function ought to return sets.

Changed in 375dd598eb2829bf993eae8eb9a41f7814ab861d.